File: results/dnd-rest-benchmark/lifecycle-runs/20260731T060905Z_claude_claude-sonnet-5_python-django/dndsite/views/characters.py
"""Stateless character math: ability modifiers, proficiency bonus, derived HP/AC."""

import json

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from dndsite.rules import ability_modifier, numeric, proficiency_bonus
from dndsite.views._util import json_body

ABILITY_KEYS = ("str", "dex", "con", "int", "wis", "cha")


@csrf_exempt
def ability_modifier_view(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        score = body["score"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    modifier = ability_modifier(score)
    if modifier is None:
        return JsonResponse({"error": "invalid score"}, status=400)

    return JsonResponse({"score": score, "modifier": modifier})


@csrf_exempt
def proficiency_view(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        level = body["level"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    bonus = proficiency_bonus(level)
    if bonus is None:
        return JsonResponse({"error": "invalid level"}, status=400)

    return JsonResponse({"level": level, "proficiency_bonus": bonus})


@csrf_exempt
def derived_stats(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        level = body["level"]
        abilities = body["abilities"]
        armor = body["armor"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    prof_bonus = proficiency_bonus(level)
    if prof_bonus is None:
        return JsonResponse({"error": "invalid level"}, status=400)

    if not isinstance(abilities, dict) or not isinstance(armor, dict):
        return JsonResponse({"error": "invalid request"}, status=400)

    modifiers = {}
    for key in ABILITY_KEYS:
        if key not in abilities:
            return JsonResponse({"error": "invalid request"}, status=400)
        mod = ability_modifier(abilities[key])
        if mod is None:
            return JsonResponse({"error": "invalid ability score"}, status=400)
        modifiers[key] = mod

    try:
        base = armor["base"]
        shield = armor["shield"]
        dex_cap = armor["dex_cap"]
    except (KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    if not isinstance(base, (int, float)) or isinstance(base, bool):
        return JsonResponse({"error": "invalid request"}, status=400)
    if not isinstance(shield, bool):
        return JsonResponse({"error": "invalid request"}, status=400)
    if not isinstance(dex_cap, (int, float)) or isinstance(dex_cap, bool):
        return JsonResponse({"error": "invalid request"}, status=400)

    hp_max = level * (6 + modifiers["con"])
    shield_bonus = 2 if shield else 0
    armor_class = numeric(base + min(modifiers["dex"], dex_cap) + shield_bonus)

    return JsonResponse(
        {
            "level": level,
            "proficiency_bonus": prof_bonus,
            "hp_max": hp_max,
            "armor_class": armor_class,
            "modifiers": modifiers,
        }
    )

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/characters.rb
# Stateless per-character calculations: ability modifiers, proficiency
# bonus, and derived combat stats (HP, AC). None of these touch the DB.

post '/v1/characters/ability-modifier' do
  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  score = body['score']
  halt 400, { error: 'invalid score' }.to_json unless integerish(score) && score.to_i.between?(1, 30)

  score = score.to_i
  { score: score, modifier: ability_modifier(score) }.to_json
end

post '/v1/characters/proficiency' do
  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  level = body['level']
  halt 400, { error: 'invalid level' }.to_json unless integerish(level) && level.to_i.between?(1, 20)

  level = level.to_i
  { level: level, proficiency_bonus: proficiency_bonus(level) }.to_json
end

post '/v1/characters/derived-stats' do
  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  level = body['level']
  abilities = body['abilities']
  armor = body['armor']

  halt 400, { error: 'invalid level' }.to_json unless integerish(level) && level.to_i.between?(1, 20)
  halt 400, { error: 'invalid abilities' }.to_json unless abilities.is_a?(Hash)
  halt 400, { error: 'invalid armor' }.to_json unless armor.is_a?(Hash)

  level = level.to_i

  %w[str dex con int wis cha].each do |key|
    score = abilities[key]
    halt 400, { error: "invalid ability #{key}" }.to_json unless integerish(score) && score.to_i.between?(1, 30)
  end

  modifiers = %w[str dex con int wis cha].each_with_object({}) do |key, acc|
    acc[key.to_sym] = ability_modifier(abilities[key].to_i)
  end

  base = armor['base']
  shield = armor['shield']
  dex_cap = armor['dex_cap']

  halt 400, { error: 'invalid armor base' }.to_json unless integerish(base)
  halt 400, { error: 'invalid armor dex_cap' }.to_json unless integerish(dex_cap)
  halt 400, { error: 'invalid armor shield' }.to_json unless shield == true || shield == false

  base = base.to_i
  dex_cap = dex_cap.to_i

  shield_bonus = shield ? 2 : 0
  hp_max = level * (6 + modifiers[:con])
  armor_class = base + [modifiers[:dex], dex_cap].min + shield_bonus

  {
    level: level,
    proficiency_bonus: proficiency_bonus(level),
    hp_max: hp_max,
    armor_class: armor_class,
    modifiers: modifiers
  }.to_json
end
